### PR TITLE
Upgrade to Scala 2.11 and Add cross build

### PR DESCRIPTION
**Pull request by helena**

### CODE REVIEW

**Comment by helena** [comment]
For anyone wanting to test aside from our internal QA, https://datastax-oss.atlassian.net/browse/SPARKC-22

File: doc/12_building_and_artifacts.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+The library package jars will be generated to:
+  - `spark-cassandra-connector/target/scala-{binary.version}/`
+  - `spark-cassandra-connector-java/target/scala-{binary.version}/`
+
+The documentation will be generated to:
+  - `spark-cassandra-connector/target/scala-{binary.version}/api/`
+  - `spark-cassandra-connector-java/target/scala-{binary.version}/api/`
+
+##### Build Tasks
+The easiest way to do this is to make the assembled connector jar using
```
Comment (RussellSpitzer) [inline doc/12_building_and_artifacts.md]: I would like to keep this in the quick use section just to show people how to get started with the Shell as quickly as possible but I definitely see the rational for moving it.

Comment (helena) [inline doc/12_building_and_artifacts.md]: I get it but we had duplicate build/artifact content, and over time I don't want to have us maintaining both. This way it is a link in both README and getting started documentation pages, where it is currently duplicated.

File: project/Settings.scala [hunk 2]
```diff
@@ -138,12 +198,16 @@ object Settings extends Build {
     }
   )
 
-  /* By default, assembly is not enabled for the demos module, but it can be enabled with
-    `-Ddemos.assembly=true`. From the command line this would be:
-     sbt -Ddemos.assembly=true twitter/assembly */
-  lazy val demoAssemblySettings =
-    if (System.getProperty("demos.assembly", "true").toBoolean) sbtAssemblySettings
-    else Seq.empty
+  lazy val theme = PromptTheme(List(
```
Comment (RussellSpitzer) [inline project/Settings.scala]: With style :)

File: spark-cassandra-connector/scala-2.11/src/main/scala/com/datastax/spark/connector/rdd/reader/AnyObjectFactory.scala [hunk 3]
```diff
@@ -0,0 +1,194 @@
+package com.datastax.spark.connector.rdd.reader
```
Comment (RussellSpitzer) [inline spark-cassandra-connector/scala-2.11/src/main/scala/com/datastax/spark/connector/rdd/reader/AnyObjectFactory.scala]: What is different in this class vs the 2.10 version?

File: spark-cassandra-connector/scala-2.11/src/main/scala/com/datastax/spark/connector/rdd/reader/AnyObjectFactory.scala [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+  val argCount: Int = constructorParamTypes.length
+
+  val argOffset: Int = oneIfMemberClass(javaClass)
+
+  @transient
+  lazy val constructorParamTypes: Array[Type] = {
+    val requiredParamClasses = javaConstructor.getParameterTypes
+      .drop(AnyObjectFactory.oneIfMemberClass(javaClass))
+
+    tpe.decl(termNames.CONSTRUCTOR).asTerm.alternatives.map { term =>
```
Comment (helena) [inline spark-cassandra-connector/scala-2.11/src/main/scala/com/datastax/spark/connector/rdd/reader/AnyObjectFactory.scala]: @RussellSpitzer this

File: spark-cassandra-connector/scala-2.11/src/main/scala/com/datastax/spark/connector/rdd/reader/ClassBasedRowReader.scala [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  private val tpe = implicitly[TypeTag[R]].tpe
+
+  // This must be  serialized:
+  val constructorArgConverters: Array[TypeConverter[_]] =
+    factory.constructorParamTypes.map(t => TypeConverter.forType(t))
+
+  @transient
+  private val setterTypes: Map[String, Type] = {
+    def argType(name: String) = {
+      val methodSymbol = tpe.decl(TermName(name)).asMethod
```
Comment (helena) [inline spark-cassandra-connector/scala-2.11/src/main/scala/com/datastax/spark/connector/rdd/reader/ClassBasedRowReader.scala]: @RussellSpitzer this is the other

File: project/Settings.scala (hunk 2 above)
Comment (RussellSpitzer) [inline project/Settings.scala]: This doesn't really respect default terminal colors so on my screen it's unreadable

image

Comment (helena) [inline project/Settings.scala]: Hrm, did not even consider that obviously. I'll see if the plugin can handle that even.

**Comment by RussellSpitzer** [comment]
Minor Other than my love of white text on black backgrounds this looks good,
+1 Code
+1 Unittests/It Tests

**Comment by helena** [comment]
@RussellSpitzer fixed shell prompt for user defaults.

**Comment by helena** [comment]
@pkolaczk Can you review this week? Thanks

File: spark-cassandra-connector/scala-2.11/src/main/scala/com/datastax/spark/connector/rdd/reader/AnyObjectFactory.scala (hunk 3 above)
Comment (pkolaczk) [inline spark-cassandra-connector/scala-2.11/src/main/scala/com/datastax/spark/connector/rdd/reader/AnyObjectFactory.scala]: I'd certainly like to *not* have to maintain this class separately for scala 2.10 and 2.11. Seems there is exactly *one* single line change in this file - the line to get the constructor declaration. Please extract the method for getting constructor into external object, that can be coded differently for 2.10 and 2.11.

File: spark-cassandra-connector/scala-2.11/src/main/scala/com/datastax/spark/connector/rdd/reader/AnyObjectFactory.scala (hunk 4 above)
Comment (pkolaczk) [inline spark-cassandra-connector/scala-2.11/src/main/scala/com/datastax/spark/connector/rdd/reader/AnyObjectFactory.scala]: Yes, please extract it to ReflectionUtil. 
The same applies to ClassBasedRowReader. There is no point in forking the whole class if the only thing that requires special treating is a well defined operation for getting the constructor definition.

File: spark-cassandra-connector/scala-2.11/src/main/scala/com/datastax/spark/connector/rdd/reader/ClassBasedRowReader.scala [hunk 6]
```diff
@@ -0,0 +1,145 @@
+package com.datastax.spark.connector.rdd.reader
```
Comment (pkolaczk) [inline spark-cassandra-connector/scala-2.11/src/main/scala/com/datastax/spark/connector/rdd/reader/ClassBasedRowReader.scala]: As stated above, no need to have this separated.

File: spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraRelation.scala [hunk 7]
```diff
@@ -9,7 +9,7 @@ import org.apache.spark.sql.catalyst.plans.logical.{Statistics, LeafNode}
 import org.apache.spark.sql.{StructField, catalyst}
 
 private[cassandra] case class CassandraRelation
-  (tableDef: TableDef, alias: Option[String])(@transient cc: CassandraSQLContext)
+  (tableDef: TableDef, alias: Option[String])(@transient val cc: CassandraSQLContext)
```
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraRelation.scala]: Why need val here?

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/GroupingBatchBuilder.scala [hunk 8]
```diff
@@ -78,7 +78,7 @@ class GroupingBatchBuilder[T](batchStatementBuilder: BatchStatementBuilder[T],
 
   /** Creates a statement from the given batch and cleans the batch so that it can be reused. */
   @inline
-  private def createStmtAndReleaseBatch(batch: Batch): RichStatement = {
+  protected def createStmtAndReleaseBatch(batch: Batch): RichStatement = {
```
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/GroupingBatchBuilder.scala]: Why protected?

File: spark-cassandra-connector/scala-2.11/src/main/scala/com/datastax/spark/connector/rdd/reader/AnyObjectFactory.scala (hunk 3 above)
Comment (helena) [inline spark-cassandra-connector/scala-2.11/src/main/scala/com/datastax/spark/connector/rdd/reader/AnyObjectFactory.scala]: Agreed. I brought up the idea of creating the primary trait and branch the changed functions in the 2 branches. I'll push that soon.

File: spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraRelation.scala (hunk 7 above)
Comment (helena) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraRelation.scala]: Does not compile in 2.11

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/GroupingBatchBuilder.scala (hunk 8 above)
Comment (helena) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/GroupingBatchBuilder.scala]: It broke in compile but I retested and it compiles in both now. Reverted.

**Comment by pkolaczk** [comment]
There are two separate compiles, once with picking files from 2.10 and another time with 2.11; and the rest of files is common. Am I right?
So I can see no problem here. The code for AnyObjectFactory and ClassBasedRowReader needs to be in the *common part*. For reflection, you call some ReflectionUtil object which has a common API, but different implementetation (separated into 2.10 and 2.11).

**Comment by helena** [comment]
@pkolaczk it can't be in the common part, the actual class names but all the code for the common (now abstractions) is in the shared source. The file names must be the same for the 2 builds, which is how I have it.

**Comment by helena** [comment]
I think having a shared abstract class in shared src containing all the code for those 2 but the one line difference per scala version, and the one line implementation for 2.10 and 2.11 is clean vs adding a new reflection util. Adding *Util* to me smells like poor design.

**Comment by pkolaczk** [comment]
Honestly, adding more inheritance smells like poor design more than adding an Util to me, particularly *implementation inheritance*. The main (only?) reason for an abstract class is when there are going to be 2 or more concrete implementations (or you at least predict there will be more than one in the future) to be switched at runtime, differing with something important. In this case, logically we've got only one, however non-portable. 

A widely applied and very successful pattern for dealing with lack of portability is extracting the non-portable bits (and only them) to external compatibility layer exposing a common API. This way you don't break SRP: the class is still responsible only for its core functionality, while the compatibility layer is responsible for portability and dealing with whatever differences there might be between Scala versions. Clear separation of concerns. And btw, if you don't like Utils, just call it `ScalaReflectionCompatibilityLayer` This way everybody knows what this class is for. Scala itself comes with plenty of singletons and *stateless* singletons are harmless.  :)

**Comment by helena** [comment]
Are you saying you want an object vs class?
Note that for an object you can not lazy val this: implicitly[TypeTag[T]].tpe, it has to decorate a class but I can make it a function.

**Comment by helena** [comment]
@pkolaczk updated PR.

**Comment by pkolaczk** [comment]
+1, please rebase / squash the commits, as there were a few back-and-forth moves.

**Comment by pkolaczk** [comment]
All tests green, both for 2.10 and 2.11.
sbt assembly ok for both as well.

**Comment by helena** [comment]
LGTM?

**Comment by helena** [comment]
@piotr will do. Thanks for pushing me to find a cleaner way on this.

**Comment by helena** [comment]
Closing for the rebased/squashed version for another round due to issues with java api. See https://github.com/datastax/spark-cassandra-connector/pull/573

**Pull request closed without merging**